src/components/maincontroller.js

```diff
@@ -1,930 +1,914 @@
+        });
+        return;
     }
 
-    function setSubtitleStreamIndex($scope, index, serverAddress) {
-        console.log('setSubtitleStreamIndex. index: ' + index);
+    $scope.userId = data.userId;
+    $scope.accessToken = data.accessToken;
+    $scope.serverAddress = data.serverAddress;
+    if (data.subtitleAppearance) {
+        window.subtitleAppearance = data.subtitleAppearance;
+    }
 
-        var currentSubtitleStream = $scope.mediaSource.MediaStreams.filter(function (m) {
-            return m.Index == $scope.subtitleStreamIndex && m.Type == 'Subtitle';
-        })[0];
-        var currentDeliveryMethod = currentSubtitleStream ? currentSubtitleStream.DeliveryMethod : null;
-
-        if (index == -1 || index == null) {
-            // Need to change the stream to turn off the subs
-            if (currentDeliveryMethod == 'Encode') {
-                console.log('setSubtitleStreamIndex video url change required');
-                var positionTicks = getCurrentPositionTicks($scope);
-                changeStream(positionTicks, { SubtitleStreamIndex: -1 });
-            } else {
-                $scope.subtitleStreamIndex = -1;
-                setTextTrack(null);
-            }
-            return;
-        }
+    // Report device capabilities
+    if (!window.hasReportedCapabilities) {
+        getMaxBitrate("Video").then((maxBitrate) => {
+            let capabilitiesUrl = $scope.serverAddress + "/Sessions/Capabilities/Full";
+            let deviceProfile = getDeviceProfile(maxBitrate);
 
-        var mediaStreams = $scope.PlaybackMediaSource.MediaStreams;
+            let capabilities = {
+                PlayableMediaTypes: ["Audio", "Video"],
+                SupportsPersistentIdentifier: false,
+                SupportsMediaControl: true,
+                DeviceProfile: deviceProfile
+            };
+            window.hasReportedCapabilities = true;
+            return fetchhelper.ajax({
+                url: capabilitiesUrl,
+                headers: getSecurityHeaders($scope.accessToken, $scope.userId),
+                type: 'POST',
+                data: JSON.stringify(capabilities),
+                contentType: 'application/json'
+            });
+        });
+    }
 
-        var subtitleStream = getStreamByIndex(mediaStreams, 'Subtitle', index);
+    data.options = data.options || {};
+    var cleanReceiverName = cleanName(data.receiverName || '');
+    window.deviceInfo.deviceName = cleanReceiverName || window.deviceInfo.deviceName;
+    // deviceId just needs to be unique-ish
+    window.deviceInfo.deviceId = cleanReceiverName ? btoa(cleanReceiverName) : window.deviceInfo.deviceId;
+
+    if (data.maxBitrate) {
+        window.MaxBitrate = data.maxBitrate;
+    }
+
+    // Items will have properties - Id, Name, Type, MediaType, IsFolder
+
+    var reportEventType;
+    var systemVolume = window.castReceiverContext.getSystemVolume();
+
+    if (data.command == 'PlayLast' || data.command == 'PlayNext') {
```

Comment:
Interesting, I forget if we have PlayLast as an action or not.

Reply:
I'm not quite sure either, I'll have to look into that. Haven't touched maincontroller very much cause I've always wanted to split it up and refactor it asap

Reply:
We still haven't documented the session controls either, so most of these options are completely unknown.